### PR TITLE
`class_<T>.property` to accept `std::function` and function object

### PR DESCRIPTION
**Pull request by hkleynhans**
Enable `class_<T>.property` to accept both `std::function` and
function object types.  This is useful in scenarios where you want a
statefull function object for purposes such as gathering statistics or
performing other generic functionality on each call.

### CODE REVIEW

**Comment by hkleynhans** [comment]
@brion , @wingo - Following on from the previous PRs, this patch expanding support for taking bsl::function and function objects on the embind interface. Please take a look.

File: site/source/docs/api_reference/bind.h.rst [hunk 1]
```diff
@@ -691,10 +697,15 @@ Classes
...
-         template<typename Getter, typename Setter>
+         template<typename PropertyType = internal::DeduceArgumentsTag, typename Getter, typename Setter>
          EMSCRIPTEN_ALWAYS_INLINE const class_& property(const char* fieldName, Getter getter, Setter setter) const
 
       This is a function template taking typenames ``Setter`` and ``Getter``: ``template<typename Getter, typename Setter>``
+      which declares a read-write property with the specified ``fieldName`` on the class. ``Getter`` and ``Setter`` may be either a
+      class method, a function, a ``std::function`` or a function object.  When ``Getter`` or ``Setter`` is not pointer-to-member-function,
+      it must accept an instance of the ``ClassType`` as the ``this`` argument. When ``Getter`` or ``Setter`` is a function object, the
+      property type must be specified as a template parameter as it cannot be deduced, e.g.:
+      ``myClass.addProperty<int>("myIntProperty", MyIntGetterFunctor(), MyIntSetterFunctor());``
```
Comment (wingo) [inline site/source/docs/api_reference/bind.h.rst]: Should this `addProperty` in the docs be `property` instead?

Comment (hkleynhans) [reply site/source/docs/api_reference/bind.h.rst]: Ahh, thanks!  Will fix this up.

**Comment by hkleynhans** [comment]
Thanks for the review @wingo !

**Comment by hkleynhans** [comment]
Re-based in attempt to get all tests to pass.

**Comment by hkleynhans** [comment]
@chadaustin @jgravelle-google  would you mind taking a look?

**Comment by kripken** [comment]
I don't know this code well enough to review it. @jgravelle-google , friendly ping?

**Comment by kripken** [comment]
Great, thanks @jgravelle-google!

@hkleynhans please merge in latest incoming to here, so we get a complete test run, and can land this.

**Comment by hkleynhans** [comment]
Thanks @jgravelle-google , @kripken , @wingo !  Merged in the latest changes as requested.

**Pull request merged**